inssqs/inssqs.go

```diff
@@ -0,0 +1,402 @@
+	client   sqs.API
+	name     string
+	url      *string
+	endpoint string
+
+	retryCount int
+
+	maxBatchSize      int
+	maxBatchSizeBytes int
+	workers           int
+
+	logger *inslogger.AppLogger
+}
+
+// Config represents the configuration settings required for initializing an SQS queue.
+type Config struct {
+	Region            string // AWS region where the SQS queue resides.
+	QueueName         string // Name of the SQS queue.
+	RetryCount        int    // Number of retry attempts allowed for queue operations.
+	MaxBatchSize      int    // Maximum size of a message batch.
+	MaxBatchSizeBytes int    // Maximum size of a message batch in bytes.
+	MaxWorkers        int    // Maximum number of workers for concurrent operations.
+	LogLevel          string // Log level for SQS operations.
+
+	EndpointUrl string // Endpoint URL for AWS operations.
+}
+
+func NewSQS(config Config) Interface {
+	if config.Region == "" {
+		panic(ErrRegionNotSet)
+	}
+
+	if config.QueueName == "" {
+		panic(ErrQueueNameNotSet)
+	}
+
+	config.setDefaults()
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.Background(),
+		awsconfig.WithRegion(config.Region),
+		awsconfig.WithRetryMaxAttempts(config.RetryCount),
+		awsconfig.WithRetryMode(aws.RetryModeAdaptive))
+	if err != nil {
+		panic(errors.Wrap(err, "error while loading aws sqs config"))
+	}
+
+	// set endpoint url if provided
+	if config.EndpointUrl != "" {
+		cfg.BaseEndpoint = aws.String(config.EndpointUrl)
+	}
+
+	var logger *inslogger.AppLogger
+	if config.LogLevel == "" {
+		logger = inslogger.NewNopLogger()
+	} else {
+		logger = inslogger.NewLogger(inslogger.LogLevel(config.LogLevel))
+	}
+
+	q := &queue{
+		client:            sqs.NewSQSProxy(awssqs.NewFromConfig(cfg)),
+		name:              config.QueueName,
+		retryCount:        config.RetryCount,
+		workers:           config.MaxWorkers,
+		logger:            logger,
+		maxBatchSize:      config.MaxBatchSize,
+		maxBatchSizeBytes: config.MaxBatchSizeBytes,
+	}
+
+	qUrl, err := q.getQueueUrl()
+	if err != nil {
+		panic(errors.Wrap(err, "error while getting queue url"))
+	}
+
+	q.url = qUrl
+
+	return q
+}
+
+func (c *Config) setDefaults() {
+	if c.MaxBatchSize == 0 {
+		c.MaxBatchSize = 10
+	}
+
+	if c.MaxBatchSizeBytes == 0 {
+		c.MaxBatchSizeBytes = 64 * 1024
+	}
+
+	if c.MaxWorkers == 0 {
+		c.MaxWorkers = 1
+	}
+
+	if c.RetryCount == 0 {
+		c.RetryCount = 3
+	}
+}
+
+// SendMessageBatch sends a batch of messages to an SQS queue, handling retries and respecting batch size constraints.
+//
+// Parameters:
+// - entries: A slice of SQSMessageEntry representing messages to be sent in batches to the SQS queue.
+//
+// Returns:
+// - failedEntries: A slice of SQSMessageEntry containing the messages that failed to be sent after all attempts.
+// - err: An error indicating any failure during the sending process, nil if all messages were sent successfully.
+func (q *queue) SendMessageBatch(entries []SQSMessageEntry) ([]SQSMessageEntry, error) {
```

Comment:
`SendMessageBatch` and `DeleteMessageBatch` is working with it's own concurrency limit. Meaning if `MaxWorkers=3` and `SendMessageBatch` is called 3 times concurrently we have 9 workers at the same time. This is okay when we think about our use case but this should mentioned in the docs as a warning note. I also think we should have the non concurrent versions of these functions.

Reply:
warn note is done.

actually user can call these functions by non-concurrently by not passing workers parameter. if they don't, workers will be 1 and message batches will be sent sequantial

Reply:
Yes but what if they want to send one message type concurrently and another one sequentially. I'm aware that this is not our usecase but can be helpful for others maybe. We can leave this as TODO as well btw.

Reply:
ok I understood, I'll note it as TODO

Reply:
Can you also add this note to the readme as well?